Fix backend reuse

arnested opened the pull request:
### What does this PR do?

This fixes the way backends are currently cached leading t some frontend configuration being ignored when re-using same backend (reported in #2323).

### Motivation

Not being able to reuse backends in different frontends is unexpected and the workaround results in defining additional, identical backends.

In a current project of mine I could potentially end up with 400 identical backends to work around this.

Fixes  #2323

### More

- [x] Added/updated tests
- [x] Added/updated documentation: No, this fixes a bug in internal behaviour

### Additional Notes

Currently backends are cached under a key generated by `entryPointName+providerName+frontend.Backend`. The `frontend.Backend` doesn't respect the differences present in frontends.

I add `Hash()` method to the `Frontend` struct and use that instead of just the backend name.

The Hash method uses [github.com/mitchellh/hashstructure](https://github.com/mitchellh/hashstructure) which is also used in [collector/collector.go](/containous/traefik/blob/master/collector/collector.go) (a nice feature of it is being able to ignore some fields when computing the hash).

dtomcej commented on server/server.go:
```diff
@@ -925,7 +925,7 @@ func (s *Server) loadConfig(configurations types.Configurations, globalConfigura
 					}
 				}
-				if backends[entryPointName+providerName+frontend.Backend] == nil {
+				if backends[entryPointName+providerName+frontend.Hash()] == nil {
```
Could this string be stored, and referenced, instead of being recalculated multiple times?

arnested replied:
Yes, that should be doable. I'll just need to make sure the frontend isn't changed between usages of a store string.

dtomcej replied:
I was just thinking inside the loop. Instead of calculating `.Hash()` three times.
`entryPointName+providerName+frontend.Hash()` in your PR.

arnested replied:
I hadn't checked if shady things where done to the frontend in the loop (although that would also have been a bug if it did).

Implemented in b716cb2b3c892a678d9cac77954dea11ae9439bc.

dtomcej commented on types/types.go:
```diff
@@ -188,6 +189,13 @@ type Frontend struct {
 
+// Hash returns the hash value of a Frontend struct.
+func (f *Frontend) Hash() string {
+	hash, _ := hashstructure.Hash(f, nil)
```
Should handle the `error` here. If the hash goes sideways, you are going to want to log or potentially return an error.

arnested replied:
Agree. I think I took a shortcut because from te definition of the `Hash()` it is not able to produce an error with the current definition of the `Frontend` struct. But that's not very kind on the people doing changes in the future.

I'll create fixes for this (today or tomorrow probably).

arnested replied:
Implemented in d44fde676f86f2d039b2b8e4798f6b25a5d98e3a.

arnested commented:
Thank you for the approve, @dtomcej.

GitHub has detected a conflict in `server/server_test.go` because the tests have been refactored into using testhelpers.

I have refactored my test method to use the new test helpers as well.

Fun story. The refactored test helpers have a bug in [`WithFrontend()`](https://github.com/containous/traefik/blob/master/testhelpers/config.go#L89): it will cache the frontends under the backend name (the fun part being this is the same kind og bug as I'm trying to fix in this pull request :smile:).

I had to rebase on master to not make everything completely unreadable. I introduced a new commit (cc69be54446b021ed3bdbbf11d770497d7fa70ae) that fixes the bug in the test helper. Next commit (fb9ed685b943d918e6b1e78659061b0ca88282d2) is my original test case refactored to use the new test helpers. The rest of the commits are as you reviewed them.
